packages/extension-driver-redshift/src/lib/redshiftDataSource.ts

```diff
@@ -0,0 +1,184 @@
+    bindParams,
+    profileName,
+    operations,
+  }: ExecuteOptions): Promise<DataResult> {
+    this.checkProfileExist(profileName);
+    const { redshiftClient, options } = this.redshiftClientMapping.get(profileName)!;
+
+    try {
+      const sqlParams: SqlParameter[] = [];
+      bindParams.forEach((value, key) => {
+        sqlParams.push({ name: key.replace(':', ''), value: String(value) });
+      });
+
+      const builtSQL = buildSQL(sql, operations);
+      let executeStatementCommandParams: ExecuteStatementCommandInput = {
+        Sql: builtSQL,
+        Database: options!.Database,
+        WorkgroupName: options!.WorkgroupName,
+      };
+      if (sqlParams.length) {
+        executeStatementCommandParams = {...executeStatementCommandParams, Parameters: sqlParams}
+      }
+
+      const executeStatementCommand = new ExecuteStatementCommand(executeStatementCommandParams);
+      const statementCommandResult = await redshiftClient.send(executeStatementCommand);
+      return await this.getResultFromExecuteStatement(statementCommandResult, redshiftClient);
+    } catch (e: any) {
+      this.logger.debug(
+        `Errors occurred, release connection from ${profileName}`
+      );
+      redshiftClient.destroy();
+      throw e;
+    }
+  }
+
+  public async prepare({ parameterIndex }: RequestParameter) {
+    // see the section of Running SQL statements with parameters when calling the Amazon Redshift Data API
+    // https://docs.aws.amazon.com/redshift/latest/mgmt/data-api.html
+    return `:${parameterIndex}`;
+  }
+
+  private async testConnection(profileName: string): Promise<DataResult | undefined> {
+    const { redshiftClient, options } = this.redshiftClientMapping.get(profileName)!; 
+    const executeStatementCommandParams: ExecuteStatementCommandInput = {
+      Sql: 'select 1',
+      Database: options!.Database,
+      WorkgroupName: options!.WorkgroupName,
+    };
+
+    const executeStatementCommand = new ExecuteStatementCommand(executeStatementCommandParams);
+
+    try {
+      const statementCommandResult = await redshiftClient.send(executeStatementCommand);
+      return await this.getResultFromExecuteStatement(statementCommandResult, redshiftClient);
+    } catch (e) {
+      redshiftClient.destroy();
+      throw e;
+    }
+  }
+
+  private async getResultFromExecuteStatement(
+    statementCommandResult: ExecuteStatementCommandOutput,
+    redshiftClient: RedshiftDataClient
+  ): Promise<DataResult> {
+    let describeStatementResponse: DescribeStatementResponse | undefined;
+    const describeStatementRequestInput: DescribeStatementCommandInput = {
+      Id: statementCommandResult.Id,
+    };
+
+    // definition of describeStatementResponse.Status
+    // https://github.com/aws/aws-sdk-js-v3/blob/29056f4ca545f7e5cf951b915bb52178305fc305/clients/client-redshift-data/src/models/models_0.ts#L604
+    while (!describeStatementResponse || describeStatementResponse.Status !== 'FINISHED') {
```

Comment:
Maybe we can add some exponential backoffs while polling status.

Reply:
I added `exp-backoff`.

Reply:
Thanks! I meant we can break between sending requests to avoid exceeding AWS API rate limiting. But I think your approach is okay, we'll retry automatically when the API returns an error.
```ts
  while (!describeStatementResponse || describeStatementResponse.Status !== 'FINISHED') {
      const describeStatementCommand = new DescribeStatementCommand(describeStatementRequestInput);
      describeStatementResponse = await redshiftClient.send(describeStatementCommand)

      if (
        describeStatementResponse.Status === 'ABORTED' || 
        describeStatementResponse.Status === 'FAILED'
      ) {
        throw describeStatementResponse.Error
      }

      // Sleep here before the next iteration
    }
```